aten/src/THC/THCCachingAllocator.cpp

```diff
@@ -587,3 +587,32 @@ THC_API uint64_t THCCachingAllocator_maxMemoryCached(int device) {
   assertValidDevice(device);
   return caching_allocator.get_stats_for_device(device).max_amount_cached;
 }
+
+//
+// CUDA IPC only allows sharing a big memory block associated with a IpcMemHandle,
+// and it can be opened only once per context per process. There are multiple
+// types of storage in the same IPC mem block, so we must cache the device ptr
+// to construct typed storage as it comes.
+// ipcMemHandle_to_devptr only saves a weak_ptr, and the shared_ptr will call
+// cudaIpcCloseMemHandle when ref_count is 0.
+//
+namespace {
+  std::mutex IpcMemMutex;
+  std::unordered_map<std::string, std::weak_ptr<void>> ipcMemHandle_to_devptr;
+}
+
+THC_API std::shared_ptr<void> THCCaching_CUDAIpcDevptr(std::string handle) {
+  std::lock_guard<std::mutex> guard(IpcMemMutex);
+  std::shared_ptr<void> sp;
+  if (ipcMemHandle_to_devptr.find(handle) == ipcMemHandle_to_devptr.end()
+      || ipcMemHandle_to_devptr[handle].expired()) {
+    void *devPtr = nullptr;
+    cudaIpcMemHandle_t ipc_handle = *(cudaIpcMemHandle_t*)handle.c_str();
+    THCudaCheck(cudaIpcOpenMemHandle(&devPtr, ipc_handle, cudaIpcMemLazyEnablePeerAccess));
+    sp = std::shared_ptr<void>(devPtr, [](void *ptr) {THCudaCheck(cudaIpcCloseMemHandle(ptr));});
+    std::weak_ptr<void> wp = sp;
+    ipcMemHandle_to_devptr[handle] = wp;
+  }
+  return ipcMemHandle_to_devptr[handle].lock();
```

Comment:
It's probably better to return sp directly inside the conditional; here, you're relying on `sp` to keep the weak pointer live so you can lock it, which is more indirect than I would like